File: C++/findBottomLeftValue.cpp
/**
 * @File    : findBottomLeftValue.cpp
 * @Brief   : 找树左下角的值
 * @Link    : https://leetcode-cn.com/problems/find-bottom-left-tree-value/
 * @Author  : Wei Li
 * @Date    : 2021-08-01
*/

#include <iostream>
#include <queue>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int findBottomLeftValue(TreeNode *root)
    {
        std::queue<TreeNode *> que;
        if (root != NULL)
            que.push(root);
        int result = 0;
        while (!que.empty())
        {
            int size = que.size();
            for (int i = 0; i < size; i++)
            {
                TreeNode *node = que.front();
                que.pop();
                if (i == 0)
                    result = node->val; // 记录最后一行第一个元素
                if (node->left)
                    que.push(node->left);
                if (node->right)
                    que.push(node->right);
            }
        }
        return result;
    }
};

int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}

File: Python/findBottomLeftValue.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findBottomLeftValue.py
@Function: 找树左下角的值
@Link: https://leetcode-cn.com/problems/find-bottom-left-tree-value/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-01
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findBottomLeftValue(self, root: TreeNode) -> int:
        depth = 0
        self.res = []

        def level(root, depth):
            if not root:
                return
            if depth == len(self.res):
                self.res.append([])
                self.res[depth].append(root.val)
            level(root.left, depth+1)
            level(root.right, depth+1)
        level(root, depth)
        return self.res[-1][0]


# ----------------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass
